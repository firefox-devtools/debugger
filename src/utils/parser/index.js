// @flow

const { workerUtils: { WorkerDispatcher } } = require("devtools-utils");

const dispatcher = new WorkerDispatcher();

const getSymbols = dispatcher.task("getSymbols");
const getVariablesInScope = dispatcher.task("getVariablesInScope");
const resolveToken = dispatcher.task("resolveToken");

module.exports = {
  getSymbols,
  getVariablesInScope,
  resolveToken,
  startParserWorker: dispatcher.start.bind(dispatcher),
  stopParserWorker: dispatcher.stop.bind(dispatcher)
};
