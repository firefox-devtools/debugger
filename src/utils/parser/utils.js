// @flow

const babylon = require("babylon");
const traverse = require("babel-traverse").default;
const t = require("babel-types");
const { isDevelopment } = require("devtools-config");
const toPairs = require("lodash/toPairs");
const get = require("lodash/get");
const isEmpty = require("lodash/isEmpty");

import type { SourceText, Location, Frame, TokenResolution } from "../../types";

const ASTs = new Map();

const symbolDeclarations = new Map();

type ASTLocation = {
  start: {
    line: number,
    column: number
  },
  end: {
    line: number,
    column: number
  }
};

export type SymbolDeclaration = {
  name: string,
  location: ASTLocation
};

export type FormattedSymbolDeclaration = {
  id: string,
  title: string,
  subtitle: string,
  value: string,
  location: ASTLocation
};

export type SymbolDeclarations = {
  functions: Array<FormattedSymbolDeclaration>,
  variables: Array<FormattedSymbolDeclaration>
};

function _parse(code) {
  return babylon.parse(code, {
    sourceType: "module",

    plugins: ["jsx", "flow"]
  });
}

function parse(text: string) {
  let ast;
  if (!text) {
    return;
  }

  try {
    ast = _parse(text);
  } catch (error) {
    if (isDevelopment()) {
      console.warn("parse failed", text);
    }

    ast = {};
  }

  return ast;
}

function getAst(sourceText: SourceText) {
  if (ASTs.has(sourceText.id)) {
    return ASTs.get(sourceText.id);
  }

  const ast = parse(sourceText.text);
  ASTs.set(sourceText.id, ast);
  return ast;
}

function getFunctionName(path) {
  if (path.node.id) {
    return path.node.id.name;
  }

  const parent = path.parent;
  if (parent.type == "ObjectProperty") {
    return parent.key.name;
  }

  if (parent.type == "ObjectExpression" || path.node.type == "ClassMethod") {
    return path.node.key.name;
  }

  if (parent.type == "VariableDeclarator") {
    return parent.id.name;
  }

  if (parent.type == "AssignmentExpression") {
    if (parent.left.type == "MemberExpression") {
      return parent.left.property.name;
    }

    return parent.left.name;
  }

  return "anonymous";
}

function isFunction(path) {
  return t.isFunction(path) ||
    t.isArrowFunctionExpression(path) ||
    t.isObjectMethod(path) ||
    t.isClassMethod(path);
}

function formatSymbol(symbol: SymbolDeclaration): FormattedSymbolDeclaration {
  return {
    id: `${symbol.name}:${symbol.location.start.line}`,
    title: symbol.name,
    subtitle: `:${symbol.location.start.line}`,
    value: symbol.name,
    location: symbol.location
  };
}

function getVariableNames(path) {
  if (t.isObjectProperty(path) && !isFunction(path.node.value)) {
    return [
      formatSymbol({
        name: path.node.key.name,
        location: path.node.loc
      })
    ];
  }

  if (!path.node.declarations) {
    return path.node.params.map(dec =>
      formatSymbol({
        name: dec.name,
        location: dec.loc
      }));
  }

  return path.node.declarations.map(dec =>
    formatSymbol({
      name: dec.id.name,
      location: dec.loc
    }));
}

function isVariable(path) {
  return t.isVariableDeclaration(path) ||
    (isFunction(path) && path.node.params.length) ||
    (t.isObjectProperty(path) && !isFunction(path.node.value));
}

function getMemberExpression(root) {
  function _getMemberExpression(node, expr) {
    if (t.isMemberExpression(node)) {
      expr = [node.property.name].concat(expr);
      return _getMemberExpression(node.object, expr);
    }

    if (t.isThisExpression(node)) {
      return ["this"].concat(expr);
    }
    return [node.name].concat(expr);
  }

  return _getMemberExpression(root, []);
}

function getSymbols(source: SourceText): SymbolDeclarations {
  if (symbolDeclarations.has(source.id)) {
    const symbols = symbolDeclarations.get(source.id);
    if (symbols) {
      return symbols;
    }
  }

  const ast = getAst(source);
  const symbols = { functions: [], variables: [] };

  traverse(ast, {
    enter(path) {
      if (isVariable(path)) {
        symbols.variables.push(...getVariableNames(path));
      }

      if (isFunction(path)) {
        symbols.functions.push(
          formatSymbol({
            name: getFunctionName(path),
            location: path.node.loc
          })
        );
      }

      if (t.isClassDeclaration(path)) {
        symbols.variables.push(
          formatSymbol({
            name: path.node.id.name,
            location: path.node.loc
          })
        );
      }
    }
  });

  symbolDeclarations.set(source.id, symbols);
  return symbols;
}

function resolveExpression(path, token: string, location: Location): ?Object {
  const node = path.node;
  if (
    t.isMemberExpression(node) &&
    node.property.name === token &&
    nodeContainsLocation({ node, location })
  ) {
    const expr = getMemberExpression(node);
    return {
      value: expr.join("."),
      location: node.loc
    };
  }

  return null;
}

function resolveScope(path, location: Location) {
  const node = path.node;
  if (
    (isFunction(path) || t.isProgram(path)) &&
    nodeContainsLocation({ node, location })
  ) {
    return path;
  }
}

// Resolves a token (at location) in the source to determine if it is in scope
// of the given frame and the expression (if any) to which it belongs
function resolveToken(
  source: SourceText,
  token: string,
  location: Location,
  frame: Frame
): ?TokenResolution {
  const ast = getAst(source);
  const scopes = [];
  let expression = null;
  let inScope = false;

  if (isEmpty(ast)) {
    return;
  }

  traverse(ast, {
    enter(path) {
      let scope = null;

      // if we haven't found an expression yet, determine if the token is part
      // of one
      if (!expression) {
        expression = resolveExpression(path, token, location);
      }

      // determine if the current path is a function or program containing the
      // frame
      scope = resolveScope(path, frame.location);
      if (scope) {
        scopes.unshift(scope);
      }
    }
  });

  // determine if the narrowest scope contains the token's location
  inScope = nodeContainsLocation({ node: scopes[0].node, location });

  return {
    expression,
    inScope
  };
}

function nodeContainsLocation({ node, location }) {
  const { start, end } = node.loc;
  const { line, column } = location;

  return !(start.line > line ||
    (start.line === line && start.column > column) ||
    end.line < line ||
    (end.line === line && end.column < column));
}

function getPathClosestToLocation(source: SourceText, location: Location) {
  const ast = getAst(source);
  let pathClosestToLocation = null;

  traverse(ast, {
    enter(path) {
      if (nodeContainsLocation({ node: path.node, location })) {
        pathClosestToLocation = path;
      }
    }
  });

  return pathClosestToLocation;
}

function getVariablesInScope(source: SourceText, location: Location) {
  const path = getPathClosestToLocation(source, location);
  const bindings = get(path, "scope.bindings", {});

  return toPairs(bindings).map(([name, binding]) => ({
    name,
    references: binding.referencePaths
  }));
}

module.exports = {
  getSymbols,
  getPathClosestToLocation,
  getVariablesInScope,
  resolveToken
};
