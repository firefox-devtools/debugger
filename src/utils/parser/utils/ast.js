// @flow

import parseScriptTags from "parse-script-tags";
import * as babylon from "babylon";
import traverse from "babel-traverse";
import isEmpty from "lodash/isEmpty";

import type { Source } from "debugger-html";

let ASTs = new Map();

function _parse(code, opts) {
  return babylon.parse(
    code,
    Object.assign({}, opts, {
      sourceType: "module",
      plugins: ["jsx", "flow", "objectRestSpread"]
    })
  );
}

function parse(text: ?string, opts?: Object) {
  let ast;
  if (!text) {
    return;
  }

  try {
    ast = _parse(text, opts);
  } catch (error) {
    ast = {};
  }

  return ast;
}

// Custom parser for parse-script-tags that adapts its input structure to
// our parser's signature
function htmlParser({ source, line }) {
  return parse(source, {
    startLine: line
  });
}

export function parseExpression(expression: string, opts?: Object) {
  return babylon.parseExpression(
    expression,
    Object.assign({}, opts, { sourceType: "script" })
  );
}

export function getAst(source: Source) {
  if (!source || !source.text) {
    return {};
  }

  if (ASTs.has(source.id)) {
    return ASTs.get(source.id);
  }

  let ast = {};
  if (source.contentType == "text/html") {
    ast = parseScriptTags(source.text, htmlParser) || {};
  } else if (source.contentType == "text/javascript") {
    ast = parse(source.text);
  }

  ASTs.set(source.id, ast);
  return ast;
}

export function clearASTs() {
  ASTs = new Map();
}

type Visitor = { enter: Function };
export function traverseAst(source: Source, visitor: Visitor) {
  const ast = getAst(source);
  if (isEmpty(ast)) {
    return null;
  }

  traverse(ast, visitor);
  return ast;
}
