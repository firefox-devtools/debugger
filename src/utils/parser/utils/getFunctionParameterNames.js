// @flow
import type { NodePath } from "babel-traverse";

export default function getFunctionParameterNames(path: NodePath): string[] {
  return path.node.params.map(param => param.name);
}
