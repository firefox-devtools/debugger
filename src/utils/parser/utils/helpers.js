// @flow

import * as t from "babel-types";
import type { NodePath, Node } from "babel-traverse";

import type { AstLocation, AstPosition } from "../types";

export function isLexicalScope(path: NodePath) {
  return t.isBlockStatement(path) || isFunction(path) || t.isProgram(path);
}

export function isFunction(path: NodePath) {
  return (
    t.isFunction(path) ||
    t.isArrowFunctionExpression(path) ||
    t.isObjectMethod(path) ||
    t.isClassMethod(path)
  );
}

export function isVariable(path: NodePath) {
  return (
    t.isVariableDeclaration(path) ||
    (isFunction(path) && path.node.params.length) ||
    (t.isObjectProperty(path) && !isFunction(path.node.value))
  );
}

export function getMemberExpression(root: Node) {
  function _getMemberExpression(node, expr) {
    if (t.isMemberExpression(node)) {
      expr = [node.property.name].concat(expr);
      return _getMemberExpression(node.object, expr);
    }

    if (t.isCallExpression(node)) {
      return [];
    }

    if (t.isThisExpression(node)) {
      return ["this"].concat(expr);
    }

    return [node.name].concat(expr);
  }

  const expr = _getMemberExpression(root, []);
  return expr.join(".");
}

export function containsPosition(a: AstLocation, b: AstPosition) {
  const startsBefore =
    a.start.line < b.line ||
    (a.start.line === b.line && a.start.column <= b.column);
  const endsAfter =
    a.end.line > b.line || (a.end.line === b.line && a.end.column >= b.column);

  return startsBefore && endsAfter;
}

export function containsLocation(a: AstLocation, b: AstLocation) {
  return containsPosition(a, b.start) && containsPosition(a, b.end);
}

export function nodeContainsPosition(node: Node, position: AstPosition) {
  const b = position.column ? position : { ...position, column: 0 };
  return containsPosition(node.loc, b);
}

export function nodeContainsLine(node: Node, position: AstPosition) {
  const a = node.loc;
  const b = position.column ? position : { ...position, column: 0 };

  console.log(a.start.line)
  const startsAtOrBefore = a.start.line === b.line;

  return startsAtOrBefore;
}
