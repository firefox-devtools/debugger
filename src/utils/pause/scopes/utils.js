/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import type { Why } from "../../../types";
import type { NamedValue } from "./types";

export function getFramePopVariables(why: Why, path: string): NamedValue[] {
  const vars: Array<NamedValue> = [];

  if (why && why.frameFinished) {
    const frameFinished = why.frameFinished;

    // Always display a `throw` property if present, even if it is falsy.
    if (Object.prototype.hasOwnProperty.call(frameFinished, "throw")) {
      vars.push({
        name: "<exception>",
        path: `${path}/<exception>`,
        contents: { value: frameFinished.throw }
      });
    }

    if (Object.prototype.hasOwnProperty.call(frameFinished, "return")) {
      const returned = frameFinished.return;

      // Do not display undefined. Do display falsy values like 0 and false. The
      // protocol grip for undefined is a JSON object: { type: "undefined" }.
      if (typeof returned !== "object" || returned.type !== "undefined") {
        vars.push({
          name: "<return>",
          path: `${path}/<return>`,
          contents: { value: returned }
        });
      }
    }
  }

  return vars;
}

export function getThisVariable(this_: any, path: string): ?NamedValue {
  if (!this_) {
    return null;
  }

  return {
    name: "<this>",
    path: `${path}/<this>`,
    contents: { value: this_ }
  };
}
