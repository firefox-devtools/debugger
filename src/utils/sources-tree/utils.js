/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import { parse } from "url";

import type { Node } from "./types";
import type { SourceRecord } from "../../types";
import { isPretty } from "../source";
const IGNORED_URLS = ["debugger eval code", "XStringBundle"];

export function nodeHasChildren(item: Node): boolean {
  return Array.isArray(item.contents);
}

export function isExactUrlMatch(pathPart: string, debuggeeUrl: string) {
  // compare to hostname with an optional 'www.' prefix
  const { host } = parse(debuggeeUrl);
  if (!host) {
    return false;
  }
  return host.replace(/^www\./, "") === pathPart.replace(/^www\./, "");
}

export function isDirectory(url: Object) {
  const parts = url.path.split("/").filter(p => p !== "");

  // Assume that all urls point to files except when they end with '/'
  // Or directory node has children
  return (
    (parts.length === 0 ||
      url.path.slice(-1) === "/" ||
      nodeHasChildren(url)) &&
    url.name != "(index)"
  );
}

export function getExtension(source: Object): string {
  const url = source.get ? source.get("url") : source.url;
  const parsedUrl = parse(url).pathname;
  if (!parsedUrl) {
    return "";
  }
  return parsedUrl.split(".").pop();
}

export function isNotJavaScript(source: Object): boolean {
  return ["css", "svg", "png"].includes(getExtension(source));
}

export function isInvalidUrl(url: Object, source: SourceRecord) {
  return (
    IGNORED_URLS.indexOf(url) != -1 ||
    !source.get("url") ||
    !url.group ||
    isPretty(source) ||
    isNotJavaScript(source)
  );
}

export function partIsFile(index: number, parts: Array<string>, url: Object) {
  const isLastPart = index === parts.length - 1;
  return !isDirectory(url) && isLastPart;
}

export function createNode(
  name: string,
  path: string,
  contents: SourceRecord | Array<Node>
): Node {
  return {
    name,
    path,
    contents
  };
}

export function createParentMap(tree: Node): WeakMap<Node, Node> {
  const map = new WeakMap();

  function _traverse(subtree) {
    if (nodeHasChildren(subtree)) {
      for (const child of subtree.contents) {
        map.set(child, subtree);
        _traverse(child);
      }
    }
  }

  // Don't link each top-level path to the "root" node because the
  // user never sees the root
  tree.contents.forEach(_traverse);
  return map;
}

export function getRelativePath(url: string) {
  const { pathname } = parse(url);
  if (!pathname) {
    return url;
  }
  const path = pathname.split("/");
  path.shift();
  return path.join("/");
}
