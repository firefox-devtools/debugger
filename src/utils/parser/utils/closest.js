// @flow

import * as t from "babel-types";

import { traverseAst } from "./ast";
import { isLexicalScope, getMemberExpression } from "./helpers";

import type { SourceText, Location } from "debugger-html";
import type { NodePath, Node } from "babel-traverse";

function getNodeValue(node: Node) {
  if (t.isThisExpression(node)) {
    return "this";
  }

  return node.name;
}

/**
 * helps find member expressions on one line and function scopes that are
 * often many lines
 */
type nodeContainsLocationParams = {
  node: Node,
  location: Location
};

function nodeContainsLocation({ node, location }: nodeContainsLocationParams) {
  const { start, end } = node.loc;
  const { line, column } = location;

  // node is a one line expression
  if (start.line === end.line) {
    return (
      start.line === line && start.column <= column && end.column >= column
    );
  }

  // node is likely a function parameter
  if (start.line === line) {
    return start.column <= column;
  }

  // node is on the same line as the closing curly
  if (end.line === line) {
    return end.column >= column;
  }

  // node is either inside the block body or outside of it
  return start.line < line && end.line > line;
}

function getClosestMemberExpression(source, token, location: Location) {
  let expression = null;
  traverseAst(source, {
    enter(path: NodePath) {
      const { node } = path;
      if (
        t.isMemberExpression(node) &&
        node.property.name === token &&
        nodeContainsLocation({ node, location })
      ) {
        const memberExpression = getMemberExpression(node);
        expression = {
          value: memberExpression,
          location: node.loc
        };
      }
    }
  });

  return expression;
}

export function getClosestExpression(
  source: SourceText,
  token: string,
  location: Location
) {
  const memberExpression = getClosestMemberExpression(source, token, location);
  if (memberExpression) {
    return memberExpression;
  }

  const path = getClosestPath(source, location);
  if (!path || !path.node) {
    return;
  }

  const { node } = path;
  return { value: getNodeValue(node), location: node.loc };
}

export function getClosestScope(source: SourceText, location: Location) {
  let closestPath = null;

  traverseAst(source, {
    enter(path) {
      if (
        isLexicalScope(path) &&
        nodeContainsLocation({ node: path.node, location })
      ) {
        closestPath = path;
      }
    }
  });

  if (!closestPath) {
    return;
  }

  return closestPath.scope;
}

export function getClosestPath(source: SourceText, location: Location) {
  let closestPath = null;

  traverseAst(source, {
    enter(path) {
      if (nodeContainsLocation({ node: path.node, location })) {
        closestPath = path;
      }
    }
  });

  return closestPath;
}
