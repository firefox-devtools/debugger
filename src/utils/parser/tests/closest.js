/* eslint max-nested-callbacks: ["error", 4]*/

import { getClosestScope, getClosestExpression } from "../utils/closest";

import { getSourceText } from "./helpers";

describe("parser", () => {
  describe("getClosestExpression", () => {
    describe("member expressions", () => {
      it("Can find a member expression", () => {
        const expression = getClosestExpression(
          getSourceText("resolveToken"),
          "x",
          {
            line: 15,
            column: 31
          }
        );

        expect(expression).toMatchSnapshot();
      });

      it("find a nested expression", () => {
        const expression = getClosestExpression(
          getSourceText("expression"),
          "secondProperty",
          {
            line: 4,
            column: 22
          }
        );

        expect(expression).toMatchSnapshot();
      });

      it("finds an expression with a call", () => {
        const expression = getClosestExpression(
          getSourceText("expression"),
          "secondProperty",
          {
            line: 6,
            column: 32
          }
        );

        expect(expression).toMatchSnapshot();
      });
    });

    it("Can find a local var", () => {
      const expression = getClosestExpression(
        getSourceText("resolveToken"),
        "beta",
        {
          line: 15,
          column: 21
        }
      );

      expect(expression).toMatchSnapshot();
    });
  });

  describe("getClosestScope", () => {
    it("finds the scope at the beginning", () => {
      const scope = getClosestScope(getSourceText("func"), {
        line: 5,
        column: 8
      });

      const node = scope.block;
      expect(node).toMatchSnapshot();
    });

    it("finds a scope given at the end", () => {
      const scope = getClosestScope(getSourceText("func"), {
        line: 9,
        column: 1
      });

      const node = scope.block;
      expect(node).toMatchSnapshot();
    });

    it("Can find the function declaration for square", () => {
      const scope = getClosestScope(getSourceText("func"), {
        line: 1,
        column: 1
      });

      const node = scope.block;
      expect(node).toMatchSnapshot();
    });
  });
});
