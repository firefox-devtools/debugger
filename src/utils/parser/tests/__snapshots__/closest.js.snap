// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser getClosestExpression Can find a local var 1`] = `
Object {
  "location": SourceLocation {
    "end": Position {
      "column": 23,
      "line": 15,
    },
    "identifierName": "beta",
    "start": Position {
      "column": 19,
      "line": 15,
    },
  },
  "value": "beta",
}
`;

exports[`parser getClosestExpression member expressions Can find a member expression 1`] = `
Object {
  "location": SourceLocation {
    "end": Position {
      "column": 31,
      "line": 15,
    },
    "start": Position {
      "column": 26,
      "line": 15,
    },
  },
  "value": "obj.x",
}
`;

exports[`parser getClosestExpression member expressions find a nested expression 1`] = `
Object {
  "location": SourceLocation {
    "end": Position {
      "column": 35,
      "line": 4,
    },
    "start": Position {
      "column": 14,
      "line": 4,
    },
  },
  "value": "obj2.c.secondProperty",
}
`;

exports[`parser getClosestExpression member expressions finds an expression with a call 1`] = `
Object {
  "location": SourceLocation {
    "end": Position {
      "column": 45,
      "line": 6,
    },
    "start": Position {
      "column": 15,
      "line": 6,
    },
  },
  "value": ".c.secondProperty",
}
`;

exports[`parser getClosestScope Can find the function declaration for square 1`] = `
FunctionDeclaration (1:0,3:1)
  async: false
  body: BlockStatement (1:19,3:1)
    body:
      - ReturnStatement (2:2,2:15)
          argument: BinaryExpression (2:9,2:14)
            left: Identifier (2:9,2:10)
              name: 'n'
            operator: '*'
            right: Identifier (2:13,2:14)
              name: 'n'
    directives: []
  expression: false
  generator: false
  id: Identifier (1:9,1:15)
    name: 'square'
  params:
    - Identifier (1:16,1:17)
        name: 'n'
`;

exports[`parser getClosestScope finds a scope given at the end 1`] = `
FunctionExpression (7:1,9:1)
  async: false
  body: BlockStatement (7:12,9:1)
    body:
      - ExpressionStatement (8:2,8:4)
          expression: NumericLiteral (8:2,8:3)
            extra:
              raw: '2'
              rawValue: 2
            value: 2
    directives: []
  expression: false
  extra:
    parenStart: 64
    parenthesized: true
  generator: false
  id: null
  params: []
`;

exports[`parser getClosestScope finds the scope at the beginning 1`] = `
FunctionExpression (5:8,5:21)
  async: false
  body: BlockStatement (5:19,5:21)
    body: []
    directives: []
  expression: false
  generator: false
  id: null
  params: []
`;
